### PR TITLE
Initial async implementation

### PR DESCRIPTION
**Pull request by shrey-garg**

### CODE REVIEW

File: container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
+package com.flipkart.poseidon.async;
```
Comment (pmohankumar) [inline container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java]: Copyright text

File: container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+import static com.flipkart.poseidon.constants.RequestConstants.BODY;
+import static com.flipkart.poseidon.constants.RequestConstants.METHOD;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * Created by shrey.garg on 29/09/18.
+ */
+public abstract class PoseidonConsumer {
+    private static final Logger logger = getLogger(PoseidonServlet.class);
```
Comment (pmohankumar) [inline container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java]: PoseidonConsumer.class

File: container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    private final Application application;
+    private final Configuration configuration;
+
+    public PoseidonConsumer(Application application, Configuration configuration) {
+        this.application = application;
+        this.configuration = configuration;
+    }
+
+    public final AsyncConsumerResult consume(AsyncConsumerRequest consumerRequest) {
+        PoseidonRequest request = new PoseidonAsyncRequest(consumerRequest.getUrl(), new HashMap<>(), new HashMap<>(), consumerRequest.getParameters());
```
Comment (pmohankumar) [inline container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java]: Though we create an immutable map copy later, let's use singleton map or something like that?

File: container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            } else if (consumerRequest.getPayload() instanceof String) {
+                requestBodyString = (String) consumerRequest.getPayload();
+            } else {
+                requestBodyString = configuration.getObjectMapper().writeValueAsString(consumerRequest.getPayload());
+            }
+
+            request.setAttribute(BODY, requestBodyString);
+        } catch (Throwable throwable) {
+            logger.error("Exception while processing async request", throwable);
+            return new AsyncConsumerResult(AsyncResultState.SIDELINE);
```
Comment (pmohankumar) [inline container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java]: What's the difference between SIDELINE and FAILURE?

File: container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        request.setAttribute(METHOD, consumerRequest.getHttpMethod());
+
+        try {
+            final String requestBodyString;
+            if (consumerRequest.getPayload() == null) {
+                requestBodyString = null;
+            } else if (consumerRequest.getPayload() instanceof String) {
+                requestBodyString = (String) consumerRequest.getPayload();
+            } else {
+                requestBodyString = configuration.getObjectMapper().writeValueAsString(consumerRequest.getPayload());
```
Comment (pmohankumar) [inline container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java]: Can we avoid a ser-deser here by adopting ParamValidationFilter to handle objects in body?

File: container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            request.setAttribute(BODY, requestBodyString);
+        } catch (Throwable throwable) {
+            logger.error("Exception while processing async request", throwable);
+            return new AsyncConsumerResult(AsyncResultState.SIDELINE);
+        }
+
+        try {
+            PoseidonResponse response = new PoseidonResponse();
+            this.application.handleRequest(request, response);
+            if (response.getStatusCode() / 100 == 2) {
```
Comment (pmohankumar) [inline container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java]: We will never require the response from PoseidonResponse in async flows?

File: container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        this.configuration = configuration;
+    }
+
+    public final AsyncConsumerResult consume(AsyncConsumerRequest consumerRequest) {
+        PoseidonRequest request = new PoseidonAsyncRequest(consumerRequest.getUrl(), new HashMap<>(), new HashMap<>(), consumerRequest.getParameters());
+        request.setAttribute(METHOD, consumerRequest.getHttpMethod());
+
+        try {
+            final String requestBodyString;
+            if (consumerRequest.getPayload() == null) {
```
Comment (pmohankumar) [inline container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java]: Lets have UT for this class

File: container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java (hunk 1 above)
Comment (shrey-garg) [reply container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java]: added in #143

File: container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java (hunk 4 above)
Comment (shrey-garg) [reply container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java]: failures are supposed to be retried

File: container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java (hunk 5 above)
Comment (shrey-garg) [reply container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java]: yes thats a TODO, just thinking if it'll affect anything else

File: container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java (hunk 6 above)
Comment (shrey-garg) [reply container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java]: probably not, but we can always change the Result class to accommodate use cases

File: container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java (hunk 3 above)
Comment (shrey-garg) [reply container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java]: used empty map

File: container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java (hunk 2 above)
Comment (shrey-garg) [reply container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java]: fixed

File: container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java (hunk 5 above)
Comment (shrey-garg) [reply container/src/main/java/com/flipkart/poseidon/async/PoseidonConsumer.java]: this is also done

**Comment by shrey-garg** [comment]
CHANGELOG updated in #143

File: api/src/main/java/com/flipkart/poseidon/internal/ParamValidationFilter.java [hunk 8]
```diff
@@ -116,6 +118,16 @@ public void filterResponse(Request request, Response response) throws InternalEr
                         } catch (IOException e) {
                             logger.error("Error in reading body : {}", e.getMessage());
                         }
+                    } else if (bodyBytes != null && bodyBytes.length > 0) {
```
Comment (pmohankumar) [inline api/src/main/java/com/flipkart/poseidon/internal/ParamValidationFilter.java]: Can we update UT for this new body param as bytes?

Comment (shrey-garg) [reply api/src/main/java/com/flipkart/poseidon/internal/ParamValidationFilter.java]: done

**Pull request merged**